numpy/core/src/umath/ufunc_object.c

```diff
@@ -4390,20 +4393,8 @@ ufunc_generic_call(PyUFuncObject *ufunc, PyObject *args, PyObject *kwds)
         return override;
     }
 
-    /*
-     * Initialize all array objects to NULL to make cleanup easier
-     * if something goes wrong.
-     */
-    for (i = 0; i < ufunc->nargs; i++) {
-        mps[i] = NULL;
-    }
-
     errval = PyUFunc_GenericFunction(ufunc, args, kwds, mps);
     if (errval < 0) {
-        for (i = 0; i < ufunc->nargs; i++) {
-            PyArray_DiscardWritebackIfCopy(mps[i]);
```

Comment:
Why is this no longer needed?

Reply:
It was impossible for it to ever be executed in the first place: the failure paths of `PyUFunc_GenericFunction` and `Py_UFuncGeneralizedFunction` both set `op` (here called `mps`) to NULL (after xdecref'ing).

But I guess it is possible that this was in fact a latent bug... @mattip, since you put the lines there, should they be moved up to the failure paths of the functions where the actual xdecref's are done? And, if so, is there a test that could be added to ensure this actually works as intended?

Reply:
Adding a line
```
if (mps[0] != NULL) printf("\nufunc is %s\n", ufunc->name);
```
and running tests yeilds over 200 times this happens in tests, for comparison ufuncs (`equal`, `greater`, ...) but also for `add`, `inner1d`, and `matrix_multiply`

On the other hand, I do not immediately remember why we would need the `PyArray_DiscardWritebackIfCopy`, but I vaguely remember a failing test that I tracked down to this call. Since then perhaps the logic has changed, I will have to go back and review where I put it in. In looking over this code, I see a few places more places with perhaps excessive use of `PyArray_DiscardWritebackIfCopy` and `NpyIter_Close`, I will go back and review it again.

Reply:
@mattip - it is not guaranteed that `mps[i] == NULL`, just that it has been properly decref'd if there is a failure. The question is if we need the `DiscardWriteBackIfCopy` in the function call (for cases, I guess, where there is a failure after the iterator is allocated).